### PR TITLE
Add comprehensive prompt testing framework

### PR DESCRIPTION
**Pull request by mattgodbolt**
Add comprehensive prompt testing framework

This PR introduces a complete testing and evaluation framework for the Claude explain
service, enabling systematic A/B testing and data-driven prompt optimization.

### Test Infrastructure
- CLI with commands: run, list, review, analyze, and improve
- YAML-based test cases covering basic optimizations, complex transformations, and edge cases
- YAML-based prompt templates with system/user/assistant sections and variable substitution
- Automatic test result storage and analysis

### Evaluation Methods
- **Automatic scoring**: Fast regex-based evaluation across multiple dimensions
  - Topic coverage detection, technical accuracy checks, clarity analysis
  - Weighted scoring: accuracy (25%), technical accuracy (25%), clarity (20%), 
    completeness (15%), length (10%), consistency (5%)
- **Claude-based AI scoring**: Deep evaluation using advanced models
  - Nuanced assessment across 5 dimensions with detailed feedback
  - Identifies missing topics, incorrect claims, and improvement areas
- **Hybrid scoring**: Configurable sampling combining both methods

### Prompt Improvement
- AI-powered prompt advisor analyzes test results and suggests specific improvements
- Constitutional AI approach: fast model generates, advanced model reviews
- Automated experimental prompt generation based on feedback
- Data-driven optimization workflow with measurable impact

### Architecture
- Fail-fast error propagation (no silent failures)
- Same core logic as production service for accurate testing
- Extensible design for custom evaluation criteria
- Support for human review workflows

## Implementation Details

- Converted test cases from JSON to YAML for better maintainability
- Added flexible template variable system (arch, language, compiler, etc.)
- Integrated with project's existing infrastructure (uv, pre-commit hooks)
- Comprehensive documentation with examples and best practices
- TODO roadmap for future enhancements (CE API integration, HTML templating)

## Usage Examples

```bash
# Run tests with AI scoring
uv run prompt-test run --prompt current --scorer claude

# Get improvement suggestions
uv run prompt-test improve --prompt current

# Compare prompt versions
uv run prompt-test run --prompt v1 --compare v2
```

This framework enables continuous improvement of explanation quality through
objective metrics and systematic testing.

Co-authored-by: Claude <noreply@anthropic.com>

### CODE REVIEW

File: prompt_testing/README.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+# Analyze all previous results
+uv run prompt-test analyze
+```
+
+## Directory Structure
+
+```
+prompt_testing/
+├── test_cases/           # Test case definitions (JSON files)
```
Comment (mattgodbolt) [inline prompt_testing/README.md:33]: I'd prefer we use YAML here as it's slightly less horrible for humans to write and keep up to date (and trivial to turn into JSON)

File: prompt_testing/README.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  ]
+}
+```
+
+## Evaluation Metrics
+
+### Automatic Scoring
+
+The framework automatically scores responses on:
```
Comment (mattgodbolt) [inline prompt_testing/README.md:82]: Can you explain how the framework scores these?

File: prompt_testing/README.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+scorer = CustomScorer()
+metrics = scorer.evaluate_response(response, expected_topics, difficulty, token_count)
+```
+
+## Troubleshooting
+
+### Common Issues
+
+1. **API Key Issues**: Ensure `ANTHROPIC_API_KEY` environment variable is set
+2. **Missing Test Cases**: Run `python -m prompt_testing.cli list` to see available cases
```
Comment (mattgodbolt) [inline prompt_testing/README.md:226]: Please update to use the `uv` script not python directly. Look for other missed cases too

File: prompt_testing/evaluation/reviewer.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                    </p>
+                    <p>
+                        <label>Why?<br><textarea name="preference_reason" placeholder="Explain your preference"></textarea></label>
+                    </p>
+                </div>
+            """
+
+        html += """
+            <script>
+                function saveReview(version) {
```
Comment (mattgodbolt) [inline prompt_testing/evaluation/reviewer.py:180]: This is a real implementation. Please don't write code that doesn't function.

File: prompt_testing/evaluation/reviewer.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                if case_id and review.case_id != case_id:
+                    continue
+                if prompt_version and review.prompt_version != prompt_version:
+                    continue
+
+                reviews.append(review)
+
+        return reviews
+
+    def create_review_interface(self, case_id: str, responses: dict[str, dict]) -> str:
```
Comment (mattgodbolt) [inline prompt_testing/evaluation/reviewer.py:72]: We should discuss the utility of this. It's a lot of extra string processing and HTML I inside python.

I'd rather use a real HTML templating system like pug if we go down this route.

File: prompt_testing/evaluation/scorer.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    response_time_ms: int | None = None
+
+    # Detailed feedback
+    missing_topics: list[str] | None = None
+    incorrect_claims: list[str] | None = None
+    notes: str | None = None
+
+
+class AutomaticScorer:
+    """Automatic scoring based on keyword matching and heuristics."""
```
Comment (mattgodbolt) [inline prompt_testing/evaluation/scorer.py:36]: This automatic scooter is pretty broad and unfocused. Where did you get the ideas for this? I think we need to discuss this more. It's a decent start I suppose.

File: prompt_testing/prompts/current.txt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+You are an expert in {arch} assembly code and {language}, helping users of the
```
Comment (mattgodbolt) [inline prompt_testing/prompts/current.txt:1]: I know this is our existing prompt but these variables here are not being expanded (I think?). And there's a to-do to not do this.

File: prompt_testing/prompts/current.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+Compiler Explorer website understand how their code compiles to assembly.
```
Comment (mattgodbolt) [inline prompt_testing/prompts/current.txt:2]: We are missing the system vs user prompt here. I think this needs to be a YAML document with both the system prompt and then the other prompts too.

File: prompt_testing/runner.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+            success = True
+            error = None
+
+        except Exception as e:
+            response_time_ms = int((time.time() - start_time) * 1000)
+            explanation = ""
+            input_tokens = 0
+            output_tokens = 0
+            success = False
+            error = str(e)
```
[the remark this answers is not in the document]
Comment (mattgodbolt) [reply prompt_testing/runner.py]: Nah this is ok

File: prompt_testing/README.md (hunk 3 above)
Comment (mattgodbolt) [reply prompt_testing/README.md]: This is now done

File: prompt_testing/README.md (hunk 2 above)
Comment (mattgodbolt) [reply prompt_testing/README.md]: Partly done. Need to update the README here.

File: prompt_testing/evaluation/reviewer.py (hunk 4 above)
Comment (mattgodbolt) [reply prompt_testing/evaluation/reviewer.py:180]: For future discussion

File: prompt_testing/evaluation/reviewer.py (hunk 5 above)
Comment (mattgodbolt) [reply prompt_testing/evaluation/reviewer.py:72]: For future discussion

File: prompt_testing/evaluation/scorer.py (hunk 6 above)
Comment (mattgodbolt) [reply prompt_testing/evaluation/scorer.py:37]: For refinement in a follow up PR

File: prompt_testing/prompts/current.txt (hunk 7 above)
Comment (mattgodbolt) [reply prompt_testing/prompts/current.txt]: Done

File: prompt_testing/prompts/current.txt (hunk 8 above)
Comment (mattgodbolt) [reply prompt_testing/prompts/current.txt]: Done

**Comment by mattgodbolt** [comment]
Good enough to merge and iterate on
